lib/std/Thread/RwLock.zig

```diff
@@ -246,4 +247,132 @@ pub const DefaultRwLock = struct {
         if ((state & READER_MASK == READER) and (state & IS_WRITING != 0))
             rwl.semaphore.post();
     }
+
+    /// The instantaneous value of the waiting-to-write state, for testing only.
+    fn isWaitingToWrite(rwl: *DefaultRwLock) bool {
+        const state = @atomicLoad(usize, &rwl.state, .SeqCst);
+        return (state & DefaultRwLock.WRITER_MASK) != 0;
+    }
 };
+
+test "DefaultRwLock - internal state" {
+    var rwl = DefaultRwLock{};
+
+    // The following failed prior to the fix for Issue #13163,
+    // where the WRITER flag was subtracted by the lock method.
+
+    try testing.expect(!rwl.isWaitingToWrite());
+    rwl.lock();
+    rwl.unlock();
+    try testing.expect(!rwl.isWaitingToWrite());
+}
+
+test "RwLock - smoke test" {
+    var rwl = RwLock{};
+
+    rwl.lock();
+    try testing.expect(!rwl.tryLock());
+    try testing.expect(!rwl.tryLockShared());
+    rwl.unlock();
+
+    try testing.expect(rwl.tryLock());
+    try testing.expect(!rwl.tryLock());
+    try testing.expect(!rwl.tryLockShared());
+    rwl.unlock();
+
+    rwl.lockShared();
+    try testing.expect(!rwl.tryLock());
+    try testing.expect(rwl.tryLockShared());
+    rwl.unlockShared();
+    rwl.unlockShared();
+
+    try testing.expect(rwl.tryLockShared());
+    try testing.expect(!rwl.tryLock());
+    try testing.expect(rwl.tryLockShared());
+    rwl.unlockShared();
+    rwl.unlockShared();
+
+    rwl.lock();
+    rwl.unlock();
+}
+
+test "RwLock - concurrent access" {
+    if (builtin.single_threaded)
+        return;
+
+    const num_writers: usize = 2;
+    const num_readers: usize = 4;
+    const num_writes: usize = 10000;
+
+    const Runner = struct {
+        const Self = @This();
+
+        rwl: RwLock = .{},
+        writes: usize = 0,
+        reads: std.atomic.Atomic(usize) = std.atomic.Atomic(usize).init(0),
+
+        term1: usize = 0,
+        term2: usize = 0,
+        term_sum: usize = 0,
+
+        fn reader(self: *Self) !void {
+            while (true) {
+                self.rwl.lockShared();
+                defer self.rwl.unlockShared();
+                if (self.writes >= num_writes) break;
+
+                try self.check();
+
+                _ = self.reads.fetchAdd(1, .AcqRel);
+            }
```

Comment:
This is effectively a spin-loop, relying RwLock or OS scheduler fairness to ensure the writer gets a chance for it to progress/stop. On an unfair scheduler this could deadlock. And in the CI, the constant readers could prevent the writer from acquiring the lock for an extended period of time. (just something to be aware of).

Also the read increment could be Monotonic/Relaxed as it's not required to happens-before or after any particular condition given it seems to be primarily a counter.

Reply:
With a severely unfair scheduler there is no way to test that a RwLock works with contended access. If you don't think we should have a contended test (as we have for Mutex), then it could be removed, but I don't recommend this.

Reply:
You can test contended access with an unfair scheduler as long as there is guaranteed progression if only one thread is scheduled. For mutex, there's a counter that can be reached with any thread, even under unfair scheduling. In this case however, the reader could spin forever as it doesn't tell the scheduler it's waiting for another thread + its stop condition relies on the writer thread being scheduled.

Reply:
I think I can address that concern by just adding a max read count that the reader threads check, so I'll do that.